File: Data Structures & Algorithms/eating-bananas/submission-5.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int max = 0;
        for (int pile : piles) {
            max = Math.max (max, pile);
        }
        int l = 1, r = max, result = max;
        while (l <= r) {
            int m = l + (r - l)/2;
            int time = 0;
            for (int pile : piles) {
                time += Math.ceil((double) pile/m);
            }
            if (time <= h) {
                result = m;
                r = m - 1;
            } else {
                l = m + 1;
            }
        }
        return result;
    }
}

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-5.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        result = 0
        stack = []

        for i in prices:
            if not stack or i < stack[-1]:
                stack.append(i)
            else:
                result = max(result, i - stack[-1])

        return result